### PR TITLE
LPS-182150 Save and retrieve properly the values of encrypted field

### PR DESCRIPTION
**Pull request by p-albuquerque**
Related issue: https://issues.liferay.com/browse/LPS-182150

### CODE REVIEW

**Comment by p-albuquerque** [comment]
Waiting [this](https://github.com/brianchandotcom/liferay-portal/pull/134141) be merged:

**Comment by p-albuquerque** [comment]
ci:test:sf

**Comment by p-albuquerque** [comment]
ci:test:sf

**Comment by p-albuquerque** [comment]
ci:test:relevant

**Comment by p-albuquerque** [comment]
ci:test:sf

File: portal-kernel/src/com/liferay/portal/kernel/util/PropsKeys.java [hunk 1]
```diff
@@ -1770,6 +1770,15 @@ public interface PropsKeys {
...
 		"notifications.max.events";
 
+	public static final String OBJECT_FIELD_ENCRYPTION_ALGORITHM =
+		"object.field.encryption.algorithm";
+
+	public static final String OBJECT_FIELD_ENCRYPTION_RESTRICTED =
+		"object.field.encryption.restricted";
+
+	public static final String OBJECT_FIELD_ENCRYPTION_SECRET =
+		"object.field.encryption.secret";
```
Comment (MarinhoFeliphe) [inline portal-kernel/src/com/liferay/portal/kernel/util/PropsKeys.java:1780]: @p-albuquerque rename to key.

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java [hunk 2]
```diff
@@ -583,6 +586,8 @@ public ObjectField updateCustomObjectField(
 
 		ObjectField newObjectField = (ObjectField)oldObjectField.clone();
 
+		_validateEncryptedBusinessType(
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java:589]: @p-albuquerque could you please ensure if we should validate on update a system object field?

Thanks!

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java [hunk 3]
```diff
@@ -710,6 +715,7 @@ private ObjectField _addObjectField(
 		ObjectDefinition objectDefinition =
 			_objectDefinitionPersistence.findByPrimaryKey(objectDefinitionId);
 
+		_validateEncryptedBusinessType(businessType, objectDefinitionId);
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java:718]: @p-albuquerque rename to `_validateEncrypted`

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java [hunk 4]
```diff
@@ -1004,6 +1010,37 @@ else if (objectFieldDBTypes.contains(dbType) &&
 		}
 	}
 
+	private void _validateEncryptedBusinessType(
+			String businessType, long objectDefinitionId)
+		throws PortalException {
+
+		if (!StringUtil.equals(
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java:1017]: @p-albuquerque use `Objects.equals` just to follow the pattern.

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java [hunk 5]
```diff
@@ -1004,6 +1010,37 @@ else if (objectFieldDBTypes.contains(dbType) &&
...
+
+		if (Validator.isNull(
+				GetterUtil.getString(
+					PropsUtil.get(PropsKeys.OBJECT_FIELD_ENCRYPTION_SECRET))) ||
+			Validator.isNull(
+				GetterUtil.getString(
+					PropsUtil.get(
+						PropsKeys.OBJECT_FIELD_ENCRYPTION_ALGORITHM)))) {
+
+			throw new NoSuchPropertiesException("Required property is missing");
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java:1031]: @p-albuquerque create an exception and use it in all places we need to check business rules of this business type.

````xml
<exception>ObjectFieldEncrypted</exception>
````

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java [hunk 6]
```diff
@@ -1004,6 +1010,37 @@ else if (objectFieldDBTypes.contains(dbType) &&
...
+
+		if (!StringUtil.equals(
+				businessType, ObjectFieldConstants.BUSINESS_TYPE_ENCRYPTED)) {
+
+			return;
+		}
+
+		if (Validator.isNull(
+				GetterUtil.getString(
+					PropsUtil.get(PropsKeys.OBJECT_FIELD_ENCRYPTION_SECRET))) ||
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java:1025]: @p-albuquerque use `PropsValues`

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java (hunk 5 above)
Comment (MarinhoFeliphe) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java:1031]: divide this props validation in two, one to validate the secret and another to validate the algorithm, messages:

1 - Missing property, PropsKeys.OBJECT_FIELD_ENCRYPTION_ALGORITHM
2 - Missing property, PropsKeys.OBJECT_FIELD_ENCRYPTION_SECRET

Comment (MarinhoFeliphe) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectFieldLocalServiceImpl.java:1031]: sort

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java [hunk 7]
```diff
@@ -1906,6 +1917,28 @@ private Expression<?> _getFunctionExpression(
...
+		}
+
+		byte[] bytes = Base64.decode(
+			GetterUtil.getString(
+				PropsUtil.get(PropsKeys.OBJECT_FIELD_ENCRYPTION_SECRET)));
+
+		return new SecretKeySpec(
+			bytes,
+			GetterUtil.getString(
+				PropsUtil.get(PropsKeys.OBJECT_FIELD_ENCRYPTION_ALGORITHM)));
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:1939]: @p-albuquerque use PropsValues and put inline.

also divide the exceptions.

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java [hunk 8]
```diff
@@ -2628,6 +2667,27 @@ else if (selectExpression instanceof ScalarDSLQueryAlias) {
...
+
+						try {
+							objects[i] = _encryptor.decrypt(
+								_getKeyObj(), (String)objects[i]);
+						}
+						catch (EncryptorException encryptorException) {
+							throw new RuntimeException(encryptorException);
+						}
+					}
+				}
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:2690]: @p-albuquerque

````java
if (encryptedObjectFields.isEmpty()) {
 continue;
}

ObjectField objectField = _objectFieldLocalService.fetchObjectField(objectDefinitionId, columnName);

if ((objectField == null) && 
    !objectField.compareBusinessType(ObjectFieldConstants.BUSINESS_TYPE_ENCRYPTED)) {

    continue;
}

encryptedObjectFields.remove(objectField);

try {
	objects[i] = _encryptor.decrypt(_getKeyObj(), (String)objects[i]);
}
catch (EncryptorException encryptorException) {
	throw new RuntimeException(encryptorException);
}
````

**Review by MarinhoFeliphe — CHANGES_REQUESTED** [review]

**Comment by p-albuquerque** [comment]
ci:test:sf

**Comment by p-albuquerque** [comment]
ci:test:sf

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java [hunk 9]
```diff
@@ -1906,6 +1917,23 @@ private Expression<?> _getFunctionExpression(
...
+		if (Validator.isNull(PropsValues.OBJECT_FIELD_ENCRYPTION_ALGORITHM)) {
+			throw new ObjectFieldEncryptedException(
+				"Missing property" +
+					PropsKeys.OBJECT_FIELD_ENCRYPTION_ALGORITHM);
+		}
+
+		if (Validator.isNull(PropsValues.OBJECT_FIELD_ENCRYPTION_KEY)) {
+			throw new ObjectFieldEncryptedException(
+				"Missing property" + PropsKeys.OBJECT_FIELD_ENCRYPTION_KEY);
+		}
```
[the remark this answers is not in the document]
Comment (p-albuquerque) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:1930]: Do you mean `RequiredEncryptedObjectFieldPropertyException`?

**Comment by p-albuquerque** [comment]
ci:test:sf

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (MarinhoFeliphe) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:1930]: @p-albuquerque exactly

**Comment by p-albuquerque** [comment]
ci:test:sf

**Comment by p-albuquerque** [comment]
ci:test:relevant

**Comment by p-albuquerque** [comment]
ci:test:sf

**Comment by p-albuquerque** [comment]
ci:test:relevant

**Comment by MarinhoFeliphe** [comment]
Hey @p-albuquerque take a look at my changes [here](https://github.com/p-albuquerque/liferay-portal/pull/60) and double check please,

Thanks!

**Pull request closed without merging**